### ISSUE TITLE
Deleting One Entry from Complex Field via delete_term Action

### ISSUE THREAD

**Issue by austins** [htmlburger/carbon-fields#498]
## Version

  - Carbon Fields: 2.2.0
  - WordPress: 4.9.4
  - PHP: 7.2.3

---

I have a use case where when a term is deleted, and if the post with the custom field has the this term's ID, then delete its entry from the complex field. This is to keep deleted "relations" in sync. This is done within the `delete_mycustomtaxonomy` ([delete_term](https://developer.wordpress.org/reference/hooks/delete_taxonomy/)) action.

Say I have this metadata for a complex field like this:
~~~
_mycomplexfield|title|0|0|value
_mycomplexfield|anotherfield|0|0|value
_mycomplexfield|title|1|0|value
_mycomplexfield|anotherfield|1|0|value
_mycomplexfield|title|2|0|value
_mycomplexfield|anotherfield|2|0|value
~~~

Note: the 'title' field is a Select field and options are term names of a taxonomy. Its stored value is the term ID.

I would like to delete where the entry index = 1, so that the result is:
~~~
_mycomplexfield|title|0|0|value
_mycomplexfield|anotherfield|0|0|value
_mycomplexfield|title|1|0|value (indices got reset, so 2 is now 1)
_mycomplexfield|anotherfield|1|0|value (indices got reset, so 2 is now 1)
~~~

I tried doing something like this that seemed to have worked in Carbon Fields v2.1.1 previously:
~~~
$termId = 42; // or whatever is passed from delete_term action
$mycomplexfieldData = carbon_get_post_meta($post->ID, 'mycomplexfield');

foreach ($mycomplexfieldData as $key => $fields) {
    if ($fields['title'] === $termId) // Delete entry if title is equal to the term ID
        unset($mycomplexfieldData[$key]);
}

carbon_set_post_meta($post->ID, 'mycomplexfield', array_values($mycomplexfieldData)); // Call array_values to reset the indices
~~~

Unexpectedly, when carbon_get_post_meta() gets called, it just changes the entry's title value to another term ID. **This only happens in the `delete_term` action.** It doesn't happen if you run this code outside of an action hook in a plugin--it works as expected.

Could this be a bug with calling carbon_get_post_meta() within the `delete_term` action?

Edit: meant to say carbon_get_post_meta() not carbon_set_post_meta() in the last paragraph.

**Comment by austins** [comment]
Another use case for similar clearing of entries would be clearing the item from an Association field when the related post is deleted.

**Comment by deqngeorgiev** [comment]
Hi @austins ,

I've tested your code and it works as you expected. Can you provide more information or the exact code you use with the actions and hooks?

**Comment by austins** [comment]
Hi @deqngeorgiev, thanks for taking a look.

This is the exact code. I have a custom post type named product and an independent taxonomy named product_spec_type.

Custom post type and taxonomy:
~~~
add_action('init', function () {
    register_post_type('product',
        array_merge(
            [labels], // large array of labels in actual code removed for this example
            ['public' => true, 'has_archive' => true, 'rewrite' => ['with_front' => false], 'supports' => ['title']]
        ));

    // This taxonomy isn't tied to any post type since it's just used to create the options list for the select field
    register_taxonomy('product_spec_type', null,
        array_merge(
            [labels], // large array of labels in actual code removed for this example
            ['publicly_queryable' => false, 'show_in_nav_menus' => false]
        ));
});

add_action('admin_menu', function () {
    // Add admin sub-menu link to Spec Types under Product menu
    $specTypesTitle = 'Spec Types';
    add_submenu_page('edit.php?post_type=product', $specTypesTitle, $specTypesTitle, 'manage_categories',
        'edit-tags.php?taxonomy=product_spec_type&post_type=product');
});
~~~

Container:
~~~
add_action('carbon_fields_register_fields', function () {
    Container::make('post_meta', 'Properties')
        ->where('post_type', '=', 'product')
        ->add_fields([
            Field::make('complex', 'specs', 'Specs')
                ->setup_labels(['plural_name' => 'Specs', 'singular_name' => 'Spec'])
                ->add_fields([
                    Field::make('select', 'product_spec_type', 'Spec Type')->set_required(true)
                        ->set_options('myplugin_getProductSpecTypeOptions')->set_width(20),
                    Field::make('text', 'product_spec_value', 'Value')->set_required(true)->set_width(80)
                ])
        ]);
});
~~~

Term delete action:
~~~
add_action('delete_product_spec_type', function ($termId) {
    $productsQuery = new WP_Query([
        'post_type' => 'product', 'posts_per_page' => -1,
        'meta_query' => [['key' => 'specs/product_spec_type', 'value' => $termId]]
    ]);

    $products = $productsQuery ->posts;
    if (!empty($products)) {
        foreach ($products as $product) {
            $specs = carbon_get_post_meta($product->ID, 'specs');

            foreach ($specs as $key => $fields) {
                if ($fields['product_spec_type'] == $termId)
                    unset($specs[$key]);
            }

            carbon_set_post_meta($product->ID, 'specs', array_values($specs));
        }
    }
});
~~~

Screenshot of Specs complex field:
[](https://i.imgur.com/eUTG8L4.png)

**Comment by austins** [comment]
I've got the solution now after debugging and deep diving into the code. The code not working in just the delete_term action and carbon_get_post_meta() returning the modified array are good hints to finding the problem.

Here are a couple things I found out:
1. carbon_get_post_meta() clones the field, then generates an array with its datastore. It doesn't just retrieve the postmeta directly from the DB. My `product_spect_type` select field has options set when requested via `set_options('myplugin_getProductSpecTypeOptions')`. This function gets a list of terms for a taxonomy and creates an array of options.
2. WordPress' `delete_term` action is run _after_ the term is deleted from the DB. This is stated on their docs and in `/wp-includes/taxonomy.php`. This means that when carbon_get_post_meta() is called, the select options will not contain the term that must be looked for and Carbon Fields sets the value to the first one in the array.

I see why Carbon Fields has to clone the field. The problem is that I must use an earlier action such as `pre_delete_term` (before error checks) or `delete_term_taxonomy` (safe). When I was searching online, I didn't see info on these actions, so I assumed that `delete_term` was run _before_ the term is deleted because WordPress has two other actions for deleting posts: `delete_post` and `deleted_post`.

It would be great if WordPress followed their naming conventions for actions, but since they already have one nearly before deletion for the term_taxonomy, that works.

Working code:
~~~
add_action('delete_term_taxonomy', function ($termTaxId) {
    $term = get_term_by('term_taxonomy_id', $termTaxId);

    if ($term->taxonomy === 'product_spec_type') {
        $productsQuery = new WP_Query([
            'post_type' => 'product', 'posts_per_page' => -1,
            'meta_query' => [['key' => 'specs/product_spec_type', 'value' => $term->term_id]]
        ]);

        $products = $productsQuery->posts;
        if (!empty($products)) {
            foreach ($products as $product) {
                $specs = carbon_get_post_meta($product->ID, 'specs');

                foreach ($specs as $key => $fields) {
                    if ($fields['product_spec_type'] == $term->term_id)
                        unset($specs[$key]);
                }

                carbon_set_post_meta($product->ID, 'specs', array_values($specs));
            }
        }
    }
});
~~~
